### PR TITLE
Implemented Ranking Table

### PR DESCRIPTION
**Pull request by kennylee9672**
- Implemented ranking table with pagination and search bar, integrate it and charts into the ranking page.
Testing:
   - run "python manage.py migrate authtoken"
   - run "python manage.py migrate"
   - run server, click "Ranking" on top of menu bar

- Implemented  steps instruction in rate page.
Testing: 
   - run "python manage.py migrate authtoken"
   - run "python manage.py migrate"
   - run server, click "Ranking" on top of menu bar.

- Reformat function separation in multiple files: two new lines between functions.
Source: https://stackoverflow.com/questions/2953250/python-pep8-blank-lines-convention

### CODE REVIEW

File: PP_APP/server/server/settings.py [hunk 1]
```diff
@@ -35,18 +35,18 @@
...
     'django.contrib.admin',
     'django.contrib.auth',
     'django.contrib.contenttypes',
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'django_cleanup',
     'rest_framework',
     'rest_framework.authtoken',
+    'rest_framework_datatables',
```
Comment (fstamenkovic) [inline PP_APP/server/server/settings.py:48]: This is missing from _requirements.txt_ so manage.py crashes. I was able to fix it by manually installing it but please update _requirements.txt_ so it gets installed automatically.

File: PP_APP/server/frontend/templates/ranking.html [hunk 2]
```diff
@@ -1,13 +1,86 @@
...
 {% block content %}
-
 <section id="about" class="about-section bg-light">
-    <div class="container">
-
-        <!-- PAGE CONTENT GOES IN HERE -->
-
-    </div>
+   <div class="container">
+      <h1 class="h3 mb-2 text-gray-800">Quick Stats</h1>
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/templates/ranking.html:6]: The charts are empty when I looked at them; is this expected behavior when the database is auto-generated?

File: PP_APP/server/frontend/templates/ranking.html [hunk 3]
```diff
@@ -1,13 +1,86 @@
...
+                  <div class="chart-pie">
+                     <canvas id="myChart4"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the percentage of the submissions that contain plastic material.</small></p>
+               </div>
+            </div>
+         </div>
+      </div>
+      <h1 class="h3 text-gray-800" style="margin-bottom: 1em; margin-top: 2em">All Packagers</h1>
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/templates/ranking.html:49]: Pagination and search bar work fantastic!

File: PP_APP/server/frontend/static/js/charts.js [hunk 4]
```diff
@@ -41,40 +41,40 @@ $(document).ready(function () {
...
+// 		});
 
-		var myChart = new Chart(ctx2, {
+		var myChart2 = new Chart(ctx2, {
 			type: 'horizontalBar',
 			data: {
 				labels: packagersData,
 				datasets: [{
-					label: '# of Pointless Packaging',
+					label: 'Number of Submission',
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/static/js/charts.js:77]: * Submissions

File: PP_APP/server/frontend/static/js/charts.js [hunk 5]
```diff
@@ -97,7 +97,7 @@ $(document).ready(function () {
 			data: {
 				labels: monthsData,
 				datasets: [{
-					label: '# of Pointless Packaging',
+					label: 'Number of Submission',
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/static/js/charts.js:100]: as above

File: PP_APP/server/frontend/static/js/charts.js [hunk 6]
```diff
@@ -120,7 +120,7 @@ $(document).ready(function () {
 			data: {
 				labels: plasticData,
 				datasets: [{
-					label: '# of Pointless Packaging',
+					label: 'Number of Submission',
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/static/js/charts.js:123]: as above

File: PP_APP/server/frontend/templates/ranking.html [hunk 7]
```diff
@@ -1,13 +1,86 @@
...
+      <div class="row">
+         <!-- Time Chart -->
+         <div class="p-3">
+            <div class="card shadow">
+               <div class="card-body">
+                  <div class="chart-area">
+                     <canvas class="my-4 w-100 chartjs-render-monitor" id="myChart3" width="1043" height="200" style="display: block; height: 244px; width: 580px;"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the Amount of pointless packaging images' submission in this year.</small></p>
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/templates/ranking.html:17]: * Amount should not be capitalized
* images' -> images
* submission -> submissions

File: PP_APP/server/frontend/templates/ranking.html [hunk 8]
```diff
@@ -1,13 +1,86 @@
...
+      <div class="row mb-4">
+         <!-- Company Chart -->
+         <div class="col mb-6">
+            <div class="card shadow">
+               <div class="card-body">
+                  <div class="chart-area">
+                     <canvas id="myChart2"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the submission of the packagers with top scores.</small></p>
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/templates/ranking.html:31]: maybe: Shows the number of submissions of the packagers with top scores.

The chart was empty so I couldn't infer what it was supposed to represent.

File: PP_APP/server/frontend/templates/rate.html [hunk 9]
```diff
@@ -1,60 +1,92 @@
...
+         <input type="email" class="form-control" id="user_email" aria-describedby="emailHelp"
+            placeholder="Enter your email address"  required>
+         <small id="emailHelp" class="form-text text-muted">This cannot be empty.</small>
+      </div>
+      <div class="row">
+         <div class="col justify-content-center">
+            <input type="submit" id="btnSubmit" name="submit" value="Rate My Package!" class="btn btn-primary btn-lg">
+         </div>
+      </div>
+   </form>
```
Comment (fstamenkovic) [inline PP_APP/server/frontend/templates/rate.html:90]: I really, really like how this page looks right now!! Great work!!

File: PP_APP/server/frontend/templates/ranking.html (hunk 2 above)
Comment (kennylee9672) [reply PP_APP/server/frontend/templates/ranking.html:6]: you need to run:
- python manage.py migrate authtoken
- python manage.py migrate
to generate data for the DB. 
Sorry I made some assumption here. And i will update this in the PR comment.

Comment (fstamenkovic) [reply PP_APP/server/frontend/templates/ranking.html:6]: I re-ran it; it still populates the table but not the charts. I looked at the Console on Google Chrome to check if something strange happened and this is what I got: 

image

Comment (ShafiTek) [reply PP_APP/server/frontend/templates/ranking.html:6]: Hmm, it works on my end out of the box. But I recommend deleting the db.sqlite3 and then run those two migrate commands. It should work hopefully.

File: PP_APP/server/frontend/templates/rate.html [hunk 10]
```diff
@@ -1,60 +1,92 @@
...
-    <img src="" id="res_img" class="rounded mx-auto d-block" style="max-height:0px;">
-    <form id="imageUploadForm" method="post" enctype="multipart/form-data">
-        {% csrf_token %}
-        
-        <div class="row">
-            <div class="col-sm">
-                <label id="top-dropzone" class="dropzone top-dropzone" style="background: url({% static 'img/sample_top.png' %}) top left no-repeat;background-size:cover;">                    
-                    <div class="text-dropzone text-center align-middle">
-                        <h2>TOP VIEW</h2>
-                        <p>(Click OR Drag and Drop)</p>
```
Comment (ShafiTek) [inline PP_APP/server/frontend/templates/rate.html:24]: Can you change "Click OR Drag and Drop" to "Add an image"?

File: PP_APP/server/frontend/templates/rate.html [hunk 11]
```diff
@@ -1,60 +1,92 @@
...
+            <div class="card-body">
+               <h2 class="card-title">Step 1.</h2>
+               <p class="card-text">Take a picture from the TOP of your package.</p>
             </div>
-            <div class="col-sm">
-                <label id="side-dropzone" class="dropzone side-dropzone" style="background: url({% static 'img/sample_side.png' %}) top left no-repeat;background-size:cover;">
-                    <div class="text-center align-middle">
-                        <div class="text-dropzone text-align-center">
-                            <h2>LOGO VIEW</h2>
-                            <p>(Click OR Drag and Drop)</p>
```
Comment (ShafiTek) [inline PP_APP/server/frontend/templates/rate.html:34]: Same as above.

File: PP_APP/server/frontend/templates/ranking.html [hunk 12]
```diff
@@ -1,13 +1,86 @@
...
+   <script src="http://cdn.datatables.net/1.10.20/js/jquery.dataTables.min.js"></script>
+   <script src="http://cdn.datatables.net/1.10.20/js/dataTables.bootstrap4.min.js"></script>
+   <script>
+      $(document).ready(function() {
+          $('#rank').DataTable({
+              'serverSide': true,
+              'ajax': '/api/rank_data/?format=datatables',
+              'columns': [
+                  {'data': 'brand_name'},
+                  {'data': 'score'},
```
Comment (ShafiTek) [inline PP_APP/server/frontend/templates/ranking.html:75]: Change this line to:
`{'data': 'score', render: $.fn.dataTable.render.number(',', '.', 1)},`

This will round the number to just 1 decimal point. ;)

File: PP_APP/server/frontend/templates/ranking.html (hunk 2 above)
Comment (fstamenkovic) [reply PP_APP/server/frontend/templates/ranking.html:6]: It still doesn't work for me; interestingly, charts work on master.

File: PP_APP/server/frontend/templates/base.html [hunk 13]
```diff
@@ -2,149 +2,134 @@
...
-    <link href="https://fonts.googleapis.com/css2?family=Basic&display=swap" rel="stylesheet">
-
-    <!-- Bootstrap core JavaScript -->
-    <script src="{% static 'vendor/jquery/jquery.min.js' %}" defer></script>
-    <script src="{% static 'vendor/bootstrap/js/bootstrap.bundle.min.js' %}" defer></script>
-    
-    <!-- Plugin JavaScript -->
-    <script src="{% static 'vendor/jquery-easing/jquery.easing.min.js' %}" defer></script>
-
-    <!-- Chart - Bootstrap Optional Theme -->
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/base.html:30]: Why was this deleted? What was edited it base.html?

File: PP_APP/server/frontend/templates/ranking.html [hunk 14]
```diff
@@ -1,13 +1,86 @@
...
+      <div class="row">
+         <!-- Time Chart -->
+         <div class="p-3">
+            <div class="card shadow">
+               <div class="card-body">
+                  <div class="chart-area">
+                     <canvas class="my-4 w-100 chartjs-render-monitor" id="myChart3" width="1043" height="200" style="display: block; height: 244px; width: 580px;"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the amount of pointless packaging images submissions in this year.</small></p>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/ranking.html:17]: "Shows the amount of pointless packaging images submissions in this year."

Should be changed to:

"The number of pointless packaging images submitted this year."

File: PP_APP/server/frontend/templates/ranking.html [hunk 15]
```diff
@@ -1,13 +1,86 @@
...
+      <div class="row mb-4">
+         <!-- Company Chart -->
+         <div class="col mb-6">
+            <div class="card shadow">
+               <div class="card-body">
+                  <div class="chart-area">
+                     <canvas id="myChart2"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the number of submissions of the packagers with top scores.</small></p>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/ranking.html:31]: "The number of submissions for top-scoring packagers."

File: PP_APP/server/frontend/templates/ranking.html [hunk 16]
```diff
@@ -1,13 +1,86 @@
...
+         <!-- Contain Chart -->
+         <div class="col mb-6">
+            <div class="card shadow">
+               <!-- Card Body -->
+               <div class="card-body">
+                  <div class="chart-pie">
+                     <canvas id="myChart4"></canvas>
+                  </div>
+                  <hr>
+                  <p style="color:gray;"><small>Shows the percentage of the submissions that contain plastic material.</small></p>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/ranking.html:44]: "Percentage of sumbission containing plastic materials."

File: PP_APP/server/frontend/templates/ranking.html [hunk 17]
```diff
@@ -1,13 +1,86 @@
...
+         </div>
+      </div>
+      <h1 class="h3 text-gray-800" style="margin-bottom: 1em; margin-top: 2em">All Packagers</h1>
+      <div class="container">
+      <div class="row">
+         <div class="col-sm-12">
+            <table id="rank" class="table table-striped table-bordered" style="width:100%">
+               <thead>
+                  <tr>
+                     <th data-data="brand_name">Brand Name</th>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/ranking.html:56]: I would change this to just "Brand" or "Company"

File: PP_APP/server/frontend/templates/ranking.html [hunk 18]
```diff
@@ -1,13 +1,86 @@
...
+                     <th data-data="score">Score</th>
+                     <th data-data="count">Submission</th>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/ranking.html:58]: "Submissions" with an 's' not just sumbission

File: PP_APP/server/frontend/templates/rate.html [hunk 19]
```diff
@@ -1,60 +1,92 @@
...
+            <div class="card-body">
+               <h2 class="card-title">Step 3.</h2>
+               <p class="card-text">Click submit and you will receive your result and score!</p>
             </div>
-        </div>
-    </form>
-    </div>
-
-    </div>
+            <img class="card-img-top" src="{% static 'img/fig.png' %}" height="250">
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/rate.html:37]: I think the style of all 3 pictures should match. The first two have a grainy background with a large orange dashed border around it and the third doesn't. I think both styles are good, but it should be consistant.

File: PP_APP/server/frontend/templates/rate.html [hunk 20]
```diff
@@ -1,60 +1,92 @@
...
+                     <h2>LOGO VIEW</h2>
+                     <p>(Click OR Drag and Drop)</p>
+                  </div>
+               </div>
+               <input type="file" hidden id="side_img" required>
+            </label>
+         </div>
+      </div>
+      <div class="form-group">
+         <label for="exampleInputEmail1">Your Email</label>
```
Comment (rmsepic) [inline PP_APP/server/frontend/templates/rate.html:80]: Maybe put some padding here, it's close to the image upload part

File: PP_APP/server/frontend/templates/base.html (hunk 13 above)
Comment (kennylee9672) [reply PP_APP/server/frontend/templates/base.html:30]: i just replace the whole file with my own version, i added some links in the base.html for my ranking table. It just easier for me to to update this base.html,  otherwise i need to go through the file line by line. and the .html is very tedious.

File: PP_APP/server/frontend/templates/rate.html (hunk 20 above)
Comment (kennylee9672) [reply PP_APP/server/frontend/templates/rate.html:80]: do you mind make a capture and send it to me in slack? because i am not sure which part you mean?

File: PP_APP/server/frontend/templates/rate.html (hunk 19 above)
Comment (kennylee9672) [reply PP_APP/server/frontend/templates/rate.html:37]: the third is the result image, so it should be the way it is otherwise could be confusing to the user.  The orange frames for the first and second images is refer to the upload section below this step section. thats why only the first two images has the fame thing.
